File: services/tradelog/tradelog-api/src/main/java/com/example/tradelog/api/controller/DividendJournalController.java
package com.example.tradelog.api.controller;

import com.example.tradelog.api.exception.TradeLogException;
import com.example.tradelog.api.facade.JournalFacade;
import com.example.tradelog.api.service.DividendJournalService;
import com.example.tradelog.api.spec.exception.ExceptionCode;
import com.example.tradelog.api.spec.model.DividendJournalModel;
import com.example.tradelog.api.spec.model.DividendTransactionsResponse;
import com.example.tradelog.api.validator.RequestValidation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping(value = "/api/v1/dividends", produces = "application/json")
public class DividendJournalController {

    private static final Logger log = LoggerFactory.getLogger(DividendJournalController.class);

    private JournalFacade journalFacade;

    public DividendJournalController(JournalFacade journalFacade) {
        this.journalFacade = journalFacade;
    }


    /**
     * Queries the list of dividends for given symbol
     * @param symbol - ex 'CVS'
     * @return list of DividendJournalModel
     * @throws TradeLogException -
     */
    @RequestMapping(value = {"/{symbol}", "/{symbol}/"}, method = RequestMethod.GET)
    @ResponseStatus(HttpStatus.OK)
    public DividendTransactionsResponse getAllBySymbol(
            @RequestHeader("accountId") String accountId,
            @PathVariable("symbol") String symbol) throws TradeLogException {

        if (!RequestValidation.validateGetAccountAndSymbol(accountId, symbol)) {
            throw new TradeLogException(ExceptionCode.BAD_REQUEST);
        }

        return new DividendTransactionsResponse.Builder()
                .withDividendItems(journalFacade.getAllDividendsBySymbol(accountId, symbol))
                .build();
    }


    /**
     * Create DIVIDEND record
     * @param accountId - account uuid
     * @param model - DividendJournalModel
     * @return created DividendJournalModel
     * @throws TradeLogException -
     */
    @RequestMapping(value = {"", "/"}, method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.OK)
    public DividendJournalModel createNewDividendRecord(
            @RequestHeader(name = "accountId") String accountId,
            @RequestBody DividendJournalModel model) throws TradeLogException {

        if (!RequestValidation.validateCreateNewDividendRecord(accountId, model)) {
            throw new TradeLogException(ExceptionCode.BAD_REQUEST);
        }

        Optional<DividendJournalModel> optionalModel = journalFacade.createDividendRecord(model);
        if (optionalModel.isEmpty()) {
            log.error("COULD NOT CREATE FOR: " + model.getTransactionDetails().getSymbol());
            throw new TradeLogException(ExceptionCode.CREATE_DIVIDEND_FAILED);
        }
        return optionalModel.get();
    }


    /**
     * Delete DIVIDEND trade record
     * @param accountId - account uuid
     * @param transactionId -
     * @throws TradeLogException -
     */
    @RequestMapping(value = {"/{transactionId}", "/{transactionId}/"}, method = RequestMethod.DELETE)
    @ResponseStatus(HttpStatus.OK)
    public void deleteDividendRecord(
            @RequestHeader(name = "accountId") String accountId,
            @PathVariable("transactionId") String transactionId) throws TradeLogException {

        if (!RequestValidation.validateDeleteDividendRecord(accountId, transactionId)) {
            throw new TradeLogException(ExceptionCode.BAD_REQUEST);
        }

        if (!journalFacade.deleteDividendRecord(accountId, transactionId)) {
            log.error("COULD NOT DELETE FOR tID: {}", transactionId);
            throw new TradeLogException(ExceptionCode.DELETE_DIVIDEND_FAILED);
        }
    }

    //TODO: missing update dividend

}

File: services/tradelog/tradelog-api/src/main/kotlin/com/example/tradelog/api/rest/controller/DividendJournalController.kt
package com.example.tradelog.api.rest.controller

import arrow.core.getOrElse
import com.example.tradelog.api.core.facade.JournalFacade
import com.example.tradelog.api.core.service.DividendJournalService
import com.example.tradelog.api.rest.DividendJournalRestInterface
import com.example.tradelog.api.rest.converter.DividendJournalModelConverter
import com.example.tradelog.api.rest.exception.ExceptionCode
import com.example.tradelog.api.rest.exception.TradeLogException
import com.example.tradelog.api.rest.validator.RequestValidator
import com.example.tradelog.api.spec.model.TLDividendJournalDto
import com.example.tradelog.api.spec.model.TLDividendTransactionsResponse
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*

@RestController
@RequestMapping(value = ["/api/v1/dividends"],
        produces = [DividendJournalController.PROTOBUF_MEDIA_TYPE_VALUE, MediaType.APPLICATION_JSON_VALUE],
        consumes = [DividendJournalController.PROTOBUF_MEDIA_TYPE_VALUE, MediaType.APPLICATION_JSON_VALUE])
class DividendJournalController(
        private val journalFacade: JournalFacade,
        private val journalService: DividendJournalService): DividendJournalRestInterface {

    companion object {
        const val PROTOBUF_MEDIA_TYPE_VALUE = "application/x-protobuf"
    }

    override fun getAllBySymbol(accountId: UUID, portfolioId: UUID, symbol: String): TLDividendTransactionsResponse {
        if (!RequestValidator.validateGetAllBySymbol(symbol)) {
            throw TradeLogException(ExceptionCode.BAD_REQUEST)
        }

        val models = journalService.getAllBySymbol(accountId, portfolioId, symbol).orNull() ?: emptyList()
        val dtos = models.map { DividendJournalModelConverter.toDto(it) }

        return TLDividendTransactionsResponse.newBuilder()
                .addAllDividendItems(dtos)
                .build()
    }


    override fun createRecord(accountId: UUID, portfolioId: UUID, dto: TLDividendJournalDto): TLDividendJournalDto {
        if (!RequestValidator.validateCreateDividendRecord(dto)) {
            throw TradeLogException(ExceptionCode.BAD_REQUEST)
        }

        return journalFacade.createDividendRecord(DividendJournalModelConverter.toModel(dto))
                .map { DividendJournalModelConverter.toDto(it) }
                .getOrElse { throw TradeLogException(ExceptionCode.CREATE_DIVIDEND_FAILED) }
    }


    override fun editRecord(accountId: UUID, portfolioId: UUID, transactionId: UUID, dto: TLDividendJournalDto) {
        if (!RequestValidator.validateEditDividendRecord(transactionId, dto)) {
            throw TradeLogException(ExceptionCode.BAD_REQUEST)
        }

        return journalFacade.editDividendRecord(transactionId, DividendJournalModelConverter.toModel(dto))
                .getOrElse { throw TradeLogException(ExceptionCode.EDIT_DIVIDEND_FAILED) }
    }


    override fun deleteRecord(accountId: UUID, portfolioId: UUID, transactionId: UUID) {
        return journalFacade.deleteDividendRecord(accountId, transactionId)
                .getOrElse { throw TradeLogException(ExceptionCode.DELETE_DIVIDEND_FAILED) }
    }
}
